### DISCUSSION TITLE
Examples for CRUD operations using react-querybuilder

### DISCUSSION THREAD

**Discussion by syednoman84** [react-querybuilder/react-querybuilder#254] [Q&A] [answered]
Hello Everyone

Can someone please share the examples of CRUD operations using this react-querybuilder? I want to understand how can I:

1. save a new query to database
2. read and load an existing query from database and display it on the front-end in the react-querybuilder
3. edit an existing query by loading it in the react-querybuilder

I am trying to learn this using api documentation but if anything is already available will be a big help.

Thanks

**Comment by jakeboone02** [comment, marked as answer]
Yeah, the documentation should probably touch on this topic a little bit. The recommended method is to use the `formatQuery` function to convert the `query` object (the parameter passed to the `onQueryChange` callback) to a `JSON.parse`-able string that you can store in a database text column.

It's best to manage the `query` in a state variable using `useState`. That way you can just use the current value of the state variable when the user wants to save the query, and you can use the setter method to update the query from an API or preloaded variable.

[This codesandbox example](https://codesandbox.io/s/react-querybuilder-save-and-load-w6g6x?file=/src/App.tsx) should point you in the right direction.

**Reply by syednoman84** [reply]
Thanks for the quick response @jakeboone02. The example is pretty good and useful. Best!

**Reply by jakeboone02** [reply]
I'd also like to point you to the [new documentation site](https://react-querybuilder.js.org/), where the [Getting Started page has this note at the bottom](https://react-querybuilder.js.org/docs/intro#training):

> For an extended tutorial on configuration and customization of `react-querybuilder`, including information about integrating it with a backend API and advanced reporting components (grids, maps, charts, etc.), check out the course [Building Advanced Admin Reporting in React](https://www.newline.co/courses/building-advanced-admin-reporting-in-react), which is taught by one of this library's maintainers, over at [newline](https://www.newline.co/).

**Reply by syednoman84** [reply]
Thanks. One thing I wanted to highlight, while using your codesandbox example, I was getting error for missing "id" parameter on two places where the "RuleGroupType" is used. One is in queryFromAPI function and second in the initialQuery function. So I added the "id" just with a value "root" to make it work. I thought I should let you know about this.

![image](https://user-images.githubusercontent.com/24880733/147371634-53953265-c22e-4efb-b9b6-2d63d15c9705.png)

![image](https://user-images.githubusercontent.com/24880733/147371645-fcec011c-20c5-4232-961e-54ba1e502413.png)

**Reply by jakeboone02** [reply]
That should only happen if you're using v3.x. If the `react-querybuilder` dependency is set to v4.x (beta) then it shouldn't require an `id`.

**Reply by syednoman84** [reply]
Got it. Thanks
